game/mips2c/jak2_functions/ocean.cpp

```diff
@@ -0,0 +1,292 @@
+//--------------------------MIPS2C---------------------
+// clang-format off
+#include "game/mips2c/mips2c_private.h"
+#include "game/kernel/jak2/kscheme.h"
+using ::jak2::intern_from_c;
+namespace Mips2C::jak2 {
+namespace draw_large_polygon_ocean {
+struct Cache {
+  void* clip_polygon_against_negative_hyperplane; // clip-polygon-against-negative-hyperplane
+  void* clip_polygon_against_positive_hyperplane; // clip-polygon-against-positive-hyperplane
+} cache;
+
+u64 execute(void* ctxt) {
+  auto* c = (ExecutionContext*)ctxt;
+  bool bc = false;
+  u32 call_addr = 0;
+  // nop                                            // sll r0, r0, 0
+  c->daddiu(sp, sp, -8);                            // daddiu sp, sp, -8
+  c->mov64(t6, s7);                                 // or t6, s7, r0
+  c->sd(ra, 0, sp);                                 // sd ra, 0(sp)
+  c->load_symbol2(t9, cache.clip_polygon_against_positive_hyperplane);// lw t9, clip-polygon-against-positive-hyperplane(s7)
+  c->mov64(a2, t4);                                 // or a2, t4, r0
+  c->mov64(a3, t5);                                 // or a3, t5, r0
+  call_addr = c->gprs[t9].du32[0];                  // function call:
+  c->daddu(t2, a2, r0);                             // daddu t2, a2, r0
+  c->jalr(call_addr);                               // jalr ra, t9
```

Comment:
Unfortunately, this is a place that requires some manual modifications because they don't actually use a normal calling convention.

I think there are 3 functions: `render-ocean-quad` calls `draw-large-polygon-ocean`, which calls `clip-polygon-against-positive-hyperplane`. Calling `draw-large-polygon-ocean` and `clip-polygon-against-positive-hyperplane` don't use the normal calling convention. For example, this function is reading the value of `t4` and `t5`.

So the modification here is to comment out this `jalr` and replace it with
```cpp
clip_polygon_against_positive_hyperplane::execute(ctxt);
```
which is a shortcut that allows you to bypass to usual assembly stuff and just directly jump to that function, while keeping all the register values. You'll have to add
```cpp
#include "game/mips2c/jak2_functions/sky.h"
```
See `debug.cpp` for an example of something similar.